File: frontend/src/features/reviews/reviewService.js
import { reviewApi } from "./reviewApi";

/**
 * reviewService chứa các hàm nghiệp vụ phía frontend liên quan đến đánh giá sản phẩm.
 */
export const reviewService = {
    /**
     * Lấy danh sách đánh giá của một sản phẩm.
     *
     * @param productId ID của sản phẩm cần lấy đánh giá
     * @param params Tham số lọc, tìm kiếm, sắp xếp và phân trang
     *
     * @returns Dữ liệu danh sách đánh giá của sản phẩm
     */
    async getProductReviews(productId, params) {
        const response = await reviewApi.getProductReviews(productId, params); // Gọi API lấy đánh giá sản phẩm.
        return response.data; // Trả về phần data từ response để component dùng trực tiếp.
    },

    /**
     * Tạo đánh giá mới cho một sản phẩm.
     *
     * @param productId ID của sản phẩm được đánh giá
     * @param data Dữ liệu đánh giá
     *
     * @returns Kết quả tạo đánh giá từ backend
     */
    async createReview(productId, data) {
        const response = await reviewApi.createReview(productId, data); // Gọi API tạo đánh giá mới.
        return response.data; // Trả về dữ liệu backend gửi về.
    },

    /**
     * Lấy danh sách đánh giá phía Admin.
     *
     * @param params Tham số lọc, tìm kiếm, sắp xếp và phân trang
     * @returns Danh sách đánh giá cho Admin
     */
    async getAdminReviews(params) {
        const response = await reviewApi.getAdminReviews(params); // Gọi API lấy danh sách đánh giá phía Admin.
        return response.data; // Trả về dữ liệu danh sách đánh giá.
    },

    /**
     * Cập nhật trạng thái đánh dấu của một đánh giá.
     *
     * @param reviewId ID của đánh giá cần cập nhật
     * @param flag Trạng thái đánh dấu mới
     *
     * @returns Kết quả cập nhật trạng thái đánh giá
     */
    async updateReviewFlag(reviewId, flag) {
        const response = await reviewApi.updateReviewFlag(reviewId, flag); // Gọi API cập nhật flag của đánh giá.
        return response.data; // Trả về message/kết quả từ backend.
    },

    /**
     * Lấy báo cáo mức độ hài lòng theo sản phẩm.
     *
     * @param params Tham số tìm kiếm, lọc, sắp xếp và phân trang
     * @returns Danh sách báo cáo đánh giá theo sản phẩm
     */
    async getReviewReport(params) {
        const response = await reviewApi.getReviewReport(params); // Gọi API lấy báo cáo mức độ hài lòng.
        return response.data; // Trả về dữ liệu báo cáo từ backend.
    }
}

File: backend/src/main/java/com/cosmetics/ecommerce/service/ReviewService.java
package com.cosmetics.ecommerce.service;

import org.springframework.data.domain.Page;

import com.cosmetics.ecommerce.dto.ProductReviewListResponseDTO;
import com.cosmetics.ecommerce.dto.ReviewReportDTO;
import com.cosmetics.ecommerce.dto.ReviewRequestDTO;
import com.cosmetics.ecommerce.dto.ReviewResponseDTO;

/**
 * Service định nghĩa các nghiệp vụ liên quan đến đánh giá sản phẩm.
 */
public interface ReviewService {
    /**
     * Tạo đánh giá mới cho sản phẩm.
     *
     * @param userId    ID người dùng tạo đánh giá
     * @param productId ID sản phẩm được đánh giá
     * @param request   Nội dung đánh giá
     * @return Thông tin đánh giá vừa tạo
     */
    ReviewResponseDTO createReview(Integer userId, Integer productId, ReviewRequestDTO request);
    
    /**
     * Lấy danh sách đánh giá của một sản phẩm.
     *
     * @param productId ID sản phẩm cần xem đánh giá
     * @param rating    Số sao cần lọc, có thể null
     * @param verified  Trạng thái verified purchase cần lọc, có thể null
     * @param keyword   Từ khóa tìm kiếm, có thể null
     * @param page      Số trang cần lấy
     * @param size      Số lượng đánh giá trên mỗi trang
     * @param sortBy    Trường dùng để sắp xếp
     * @param sortDir   Chiều sắp xếp
     * @return Thông tin tổng hợp đánh giá sản phẩm và danh sách đánh giá theo trang
     */
    ProductReviewListResponseDTO getProductReviews(
        Integer productId, 
        Integer rating,
        Boolean verified,
        String keyword,
        int page,
        int size,
        String sortBy,
        String sortDir
    );
    
    /**
     * Lấy danh sách đánh giá dành cho Admin.
     *
     * @param rating    Số sao cần lọc, có thể null
     * @param flag      Trạng thái kiểm duyệt cần lọc, có thể null
     * @param verified  Trạng thái verified purchase cần lọc, có thể null
     * @param productId ID sản phẩm cần lọc, có thể null
     * @param keyword   Từ khóa tìm kiếm, có thể null
     * @param page      Số trang cần lấy
     * @param size      Số lượng đánh giá trên mỗi trang
     * @param sortBy    Trường dùng để sắp xếp
     * @param sortDir   Chiều sắp xếp
     * @return Danh sách đánh giá dành cho Admin theo dạng phân trang
     */
    Page<ReviewResponseDTO> getAllReviewsForAdmin(
        Integer rating,
        String flag,
        Boolean verified,
        Integer productId,
        String keyword,
        int page,
        int size,
        String sortBy,
        String sortDir
    );

    /**
     * Cập nhật trạng thái kiểm duyệt của một đánh giá.
     *
     * @param reviewId ID đánh giá cần cập nhật
     * @param flag     Trạng thái kiểm duyệt mới
     * @return Thông tin đánh giá sau khi cập nhật
     */
    ReviewResponseDTO updateReviewFlag(Integer reviewId, String flag);
    
    /**
     * Lấy báo cáo tổng hợp đánh giá theo sản phẩm.
     *
     * @param keyword          Từ khóa tìm kiếm, có thể null
     * @param minAverageRating Điểm đánh giá trung bình tối thiểu, có thể null
     * @param page             Số trang cần lấy
     * @param size             Số lượng report trên mỗi trang
     * @param sortBy           Trường dùng để sắp xếp
     * @param sortDir          Chiều sắp xếp
     * @return Báo cáo tổng hợp đánh giá theo dạng phân trang
     */
    Page<ReviewReportDTO> getReviewReport(
        String keyword,
        Double minAverageRating,
        int page,
        int size,
        String sortBy,
        String sortDir
    );
}
